### PR TITLE
fix: Updated code and version requirements to work with Terraform 0.15

### LINKED ISSUE #1164: "list" function was deprecated in Terraform v0.12
This module doesn't work with Terraform v0.15.0-dev build from source (commit c0e37a7ea1928918a9efc02290ecd241d396d3a6)

```
Error: Error in function call

  on .terraform/modules/eks/outputs.tf line 3, in output "cluster_id":
   3:   value       = element(concat(aws_eks_cluster.this.*.id, list("")), 0)

Call to function "list" failed: the "list" function was deprecated in
Terraform v0.12 and is no longer available; use tolist([ ... ]) syntax to
write a literal list.
```

https://github.com/terraform-aws-modules/terraform-aws-eks/blob/65777d0a37791dd1200d478e3bacc4c9a75f24c8/outputs.tf#L3

### PR DESCRIPTION
**Pull request by TjeuKayim**
# PR o'clock

## Description

Replaced use of [deprecated](https://www.terraform.io/docs/configuration/functions/list.html) `list()` function with `[]` syntax and `map()` function with `{}` syntax.

Fix #1164

### Checklist

- [x] CI tests are passing
- [x] README.md has been updated after any changes to variables and outputs. See https://github.com/terraform-aws-modules/terraform-aws-eks/#doc-generation

### CODE REVIEW

File: workers_launch_template.tf [hunk 1]
```diff
@@ -100,7 +100,7 @@ resource "aws_autoscaling_group" "workers_launch_template" {
 
   dynamic "mixed_instances_policy" {
     iterator = item
-    for_each = (lookup(var.worker_groups_launch_template[count.index], "override_instance_types", null) != null) || (lookup(var.worker_groups_launch_template[count.index], "on_demand_allocation_strategy", local.workers_group_defaults["on_demand_allocation_strategy"]) != null) ? list(var.worker_groups_launch_template[count.index]) : []
+    for_each = (lookup(var.worker_groups_launch_template[count.index], "override_instance_types", null) != null) || (lookup(var.worker_groups_launch_template[count.index], "on_demand_allocation_strategy", local.workers_group_defaults["on_demand_allocation_strategy"]) != null) ? toList([var.worker_groups_launch_template[count.index]]) : []
```
Comment (barryib) [inline workers_launch_template.tf:103]:
```suggestion
    for_each = (lookup(var.worker_groups_launch_template[count.index], "override_instance_types", null) != null) || (lookup(var.worker_groups_launch_template[count.index], "on_demand_allocation_strategy", local.workers_group_defaults["on_demand_allocation_strategy"]) != null) ? tolist([var.worker_groups_launch_template[count.index]]) : []
```

**Comment by TjeuKayim** [comment]
Thanks for the review @barryib. I applied your suggestion and rebased on top of master.

**Comment by TjeuKayim** [comment]
I found that the deprecated function `map()` is also going to be removed in Terraform 0.15. I'm adding a simple patch for that one too.

**Comment by TjeuKayim** [comment]
The CI "Check Docs" failure is probably a false positive as I didn't alter anything related to the documentation in this pull request.

File: workers_launch_template.tf [hunk 2]
```diff
@@ -100,7 +100,7 @@ resource "aws_autoscaling_group" "workers_launch_template" {
 
   dynamic "mixed_instances_policy" {
     iterator = item
-    for_each = (lookup(var.worker_groups_launch_template[count.index], "override_instance_types", null) != null) || (lookup(var.worker_groups_launch_template[count.index], "on_demand_allocation_strategy", local.workers_group_defaults["on_demand_allocation_strategy"]) != null) ? list(var.worker_groups_launch_template[count.index]) : []
+    for_each = (lookup(var.worker_groups_launch_template[count.index], "override_instance_types", null) != null) || (lookup(var.worker_groups_launch_template[count.index], "on_demand_allocation_strategy", local.workers_group_defaults["on_demand_allocation_strategy"]) != null) ? tolist([var.worker_groups_launch_template[count.index]]) : []
```
Comment (antonbabenko) [inline workers_launch_template.tf:103]: `tolist()` is not needed here because there is already a list provided (`[]`).

Comment (TjeuKayim) [reply workers_launch_template.tf:103]: ok, let me fix that

**Review by antonbabenko — APPROVED** [review]

**Comment by RichiCoder1** [comment]
Terraform 0.15 is out and officially breaking things 😁

**Comment by barryib** [comment]
> Terraform 0.15 is out and officially breaking things 😁

Will look into this during the week.

**Comment by bappr** [comment]
Hey guys,
When can we expect this one to be merged, we are currently being impacted by https://github.com/hashicorp/terraform/issues/27172 that has been fixed in 0.15 but can't use 0.15 because of this current PR not merged.
Thanks!

**Comment by PavelSusloparov** [comment]
bad luck for people who use docker.mirror.hashicorp.services/hashicorp/terraform:light, terraform is updated to 0.15.0 there and terraform-aws-eks deps are failing. 
Can we merge this in please?

**Comment by andyli** [comment]
It's a rather trivial change if you take a look at the file diff. Hope it wouldn't take too long to be merged.

**Comment by jtgorny** [comment]
Pls fix :sweat_smile:
`
Call to function "list" failed: the "list" function was deprecated in Terraform v0.12 and is no longer available; use tolist([ ... ]) syntax to write a literal list.
`

**Comment by antonbabenko** [comment]
Terraform 0.15 is supported in version [v15.0.0](https://github.com/terraform-aws-modules/terraform-aws-eks/releases/v15.0.0) that has been just released.

For the previous version of Terraform please use this module version `~> 14.0`.

Let us know if there are any issues with this release.

@barryib I used the same release process as for all other modules we have and CHANGELOG changes were rather big ( https://github.com/terraform-aws-modules/terraform-aws-eks/commit/eafc4582e7ecc50a45e3c283f5d8acaed9976afa )

**Comment by vauchok** [comment]
Hi @antonbabenko,

I think you've missed `list(var.worker_groups_launch_template[count.index]` in `.terraform/modules/eks/workers_launch_template.tf line 167`.

```
│ Error: Error in function call
│ 
│   on .terraform/modules/eks/workers_launch_template.tf line 167, in resource "aws_autoscaling_group" "workers_launch_template":
│  167:     for_each = (lookup(var.worker_groups_launch_template[count.index], "override_instance_types", null) != null) || (lookup(var.worker_groups_launch_template[count.index], "on_demand_allocation_strategy", local.workers_group_defaults["on_demand_allocation_strategy"]) != null) ? [] : list(var.worker_groups_launch_template[count.index])
│     ├────────────────
│     │ count.index is 1
│     │ var.worker_groups_launch_template is tuple with 4 elements
│ 
│ Call to function "list" failed: the "list" function was deprecated in
│ Terraform v0.12 and is no longer available; use tolist([ ... ]) syntax to
│ write a literal list.
```

**Comment by marcuz** [comment]
Hi @antonbabenko,

adding to @vauchok, also a `map` call is missing:

```
│ Error: Error in function call
│ 
│   on .terraform/modules/sud.eks_main/workers_launch_template.tf line 212, in resource "aws_autoscaling_group" "workers_launch_template":
│  212:         map(
│  213:           "key", tag_key,
│  214:           "value", tag_value,
│  215:           "propagate_at_launch", "true"
│  216:         )
│ 
│ Call to function "map" failed: the "map" function was deprecated in Terraform v0.12 and is no longer available; use tomap({ ... }) syntax to write a literal map.
```

**Comment by antonbabenko** [comment]
Weird that `terraform validate` command did not catch these errors. Fixed in #1307.

**Comment by antonbabenko** [comment]
Big thanks for the report!

[v15.1.0](https://github.com/terraform-aws-modules/terraform-aws-eks/releases/v15.1.0) has been just released.

**Comment by marcuz** [comment]
@antonbabenko works now, thank you very much! :raised_hands:

**Pull request merged**